src/com/perl5/lang/perl/idea/refactoring/extract/PerlExtractMethodHandler.java

```diff
@@ -0,0 +1,139 @@
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.perl5.lang.perl.idea.refactoring.extract;
+
+import com.intellij.codeInsight.codeFragment.CodeFragment;
+import com.intellij.openapi.actionSystem.DataContext;
+import com.intellij.openapi.editor.Editor;
+import com.intellij.openapi.editor.SelectionModel;
+import com.intellij.openapi.project.Project;
+import com.intellij.psi.PsiElement;
+import com.intellij.psi.PsiFile;
+import com.intellij.refactoring.RefactoringActionHandler;
+import com.intellij.refactoring.extractMethod.AbstractExtractMethodDialog;
+import com.intellij.refactoring.extractMethod.ExtractMethodDecorator;
+import com.intellij.refactoring.extractMethod.ExtractMethodValidator;
+import com.intellij.refactoring.util.AbstractVariableData;
+import com.intellij.util.containers.ArrayListSet;
+import com.perl5.lang.perl.fileTypes.PerlFileTypeScript;
+import com.perl5.lang.perl.idea.refactoring.PerlRefactorUtils;
+import com.perl5.lang.perl.idea.refactoring.PerlRefactor;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import java.util.Set;
+
+public class PerlExtractMethodHandler implements RefactoringActionHandler {
+
+  @Override
+  public void invoke(@NotNull Project project, Editor editor, PsiFile file, DataContext dataContext) {
+
+    final SelectionModel selectionModel = editor.getSelectionModel();
+    if (!selectionModel.hasSelection()) selectionModel.selectLineAtCaret();
+
+    final int selectionStartLine = selectionModel.getSelectionStart();
+    final int selectionLength = selectionModel.getSelectionEnd() - selectionStartLine;
+    String selectedText = selectionModel.getSelectedText();
+
+    if (selectedText == null || selectedText.isEmpty()) {
+      return;
+    }
+
+    PerlRefactor perlRefactor = new PerlRefactor(selectedText, "myNewMethod");
+
+    Set<String> parameterVariables = new ArrayListSet<>();
+    parameterVariables.addAll(perlRefactor.getParameters());
+
+    Set<String> returnVariables = new ArrayListSet<>();
+    returnVariables.addAll(perlRefactor.getInnerVariables());
+
+    // TODO: create PerlCodeFragment that contains my(...) = foo(bar);
+    // TODO: figure how to use returnInside
+    CodeFragment fragment = new CodeFragment(parameterVariables, returnVariables,false);
+
+    // This populates the parameters and signature
+    final ExtractMethodDecorator decorator = (methodName, variableData) -> {
+
+      StringBuilder builder = new StringBuilder();
+      boolean first = true;
+
+      if (perlRefactor.getOuterVariablesModified().size() > 0) {
+        builder.append("my (");
+        for (String ov : perlRefactor.getOuterVariablesModified()) {
+          if (first) {
```

Comment:
build a list than join it with `StringUtil.join`

Reply:
Is that faster or something ?